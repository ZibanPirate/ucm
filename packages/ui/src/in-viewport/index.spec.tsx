import { render } from "@testing-library/react";

import { InViewport, InViewportProps } from ".";

class MockedIntersectionObserver implements IntersectionObserver {
  private static staticCallback: IntersectionObserverCallback;
  private static staticObserver: IntersectionObserver;
  static changeVisibility(visible: boolean) {
    MockedIntersectionObserver.staticCallback(
      [{ isIntersecting: visible } as IntersectionObserverEntry],
      MockedIntersectionObserver.staticObserver,
    );
  }

  root!: Document | Element | null;
  rootMargin!: string;
  thresholds: readonly number[] = [];

  constructor(callback: IntersectionObserverCallback) {
    MockedIntersectionObserver.staticCallback = callback;
    MockedIntersectionObserver.staticObserver = this;
  }

  disconnect = (): null => null;
  observe = (): null => null;
  takeRecords = (): IntersectionObserverEntry[] => [];
  unobserve = (): null => null;
}

describe(`Testing component '${InViewport.name}' :`, () => {
  it(`should render and notify when it enters the viewport then notify when it leaves it`, () => {
    window.IntersectionObserver = MockedIntersectionObserver;

    const onVisibilityChanged = jest.fn();
    const { container } = render(<InViewport onVisibilityChanged={onVisibilityChanged} />);
    expect(container).toMatchSnapshot();

    MockedIntersectionObserver.changeVisibility(true);
    expect(onVisibilityChanged).toHaveBeenNthCalledWith<
      Parameters<InViewportProps["onVisibilityChanged"]>
    >(1, "entered");
    MockedIntersectionObserver.changeVisibility(false);
    expect(onVisibilityChanged).toHaveBeenNthCalledWith<
      Parameters<InViewportProps["onVisibilityChanged"]>
    >(2, "left");
  });
});
