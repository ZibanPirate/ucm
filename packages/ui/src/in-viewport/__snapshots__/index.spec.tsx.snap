// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing component 'InViewport' : should render and notify when it enters the viewport then notify when it leaves it 1`] = `
<div>
  <div />
</div>
`;
